File: StockReporter/Queries/close_price_query.rb
require_relative '../Repositories/mongo_repository'

class ClosePriceQuery

  def self.query(params)
    date = params[:date]
    share_index = params[:share_index]

    repo = MongoRepository.new("Equity")

    start_date = Time.utc(date.year, date.month, date.day, 0, 0)
    end_date = Time.utc(date.year, date.month, date.day, 23, 59)

    equity = repo.find_by_query({ :$and =>
      [
        {:ShareIndex => share_index},
        {:DateOfPrice => {:$gte => start_date}},
        {:DateOfPrice => {:$lte => end_date}},
        {:Latest => true}
      ]}).first

    return -1 if equity.nil?

    equity["ClosePrice"]
  end

end

File: stockanalyzer_elixir/apps/technicalanalysis/lib/queries/close_price_query.ex
defmodule ClosePriceQuery do
  use Timex

  def query(share, date) do

    start_date = date |> Date.shift(days: -5)

    start_date_part = {start_date.year, start_date.month, start_date.day}
    end_date_part = {date.year, date.month, date.day}

    start_time = Date.from({start_date_part, {0, 0, 0}})
    end_time = Date.from({end_date_part, {23, 59, 59}})

    item = MongoRepository.find(
                                %{
                                    '$and':
                                    [
                                    %{ShareIndex: share},
                                    %{DateOfPrice: %{'$gte': start_time |> date_to_bson }},
                                    %{DateOfPrice: %{'$lte': end_time |> date_to_bson }},
                                    %{Latest: true}]
                                },"Equity")

    |> Enum.sort(&(&1[:DateOfPrice] > &2[:DateOfPrice]))
    |> Enum.at 0

    close = item[:ClosePrice]

    case close do
      nil -> -1
      x when is_number(x) -> close
      x when is_number(x) != True -> Float.parse(close) |> elem(0)
    end

  end

  def date_to_bson(date) do
    ms = Date.to_secs(date, :epoch) * 1000
    %Bson.UTC{ms: ms}
  end

end
